theia-extensions/viewer-prototype/src/node/trace-server-service.ts

```diff
@@ -1,60 +1,88 @@
-
-    startTraceServer(path: string | undefined, port: number | undefined): Promise<string> {
-        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-        const server = spawn(path!, ['-vmargs', `-Dtraceserver.port=${port}`]);
-        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-        this.processId = server.pid!;
-        const timeouts: NodeJS.Timeout[] = [];
-
-        return new Promise<string>((resolve, reject) => {
-            // TODO: Since the resolve never actually gets returned, we need a better way to determine if the child process
-            // initiated by the spawn successfully starts the trace server
-
-            // If the server doesn't error or exit within 2 seconds, we consider that a success.
-            // That doesn't mean that the server has really started. On the frontend, the `TraceServerConnectionStatusService`
-            // will ping the port until it receives a response, which is the official measure of success.
 
-            timeouts.push(setTimeout(() => resolve('success'), 2000));
+    private server?: ChildProcess;
+    private port?: number;
 
-            // If the server exits or errors before it outputs, consider it a failure.
-            server.on('error', err => {
-                reject(err);
-            });
-            server.on('exit', code => {
-                reject(PortBusy(code));
+    async startTraceServer(path?: string, port?: number): Promise<string> {
+        if (path === undefined) {
+            throw new Error('no Trace Server path specified');
+        } else if (port === undefined) {
+            throw new Error('no Trace Server port specified');
+        } else if (this.isServerRunning(this.server)) {
+            if (this.port === port) {
+                return SUCCESS;
+            } else {
+                throw new Error('the Trace Server is already running on a different port');
+            }
+        }
+        const server = spawn(path, ['-vmargs', `-Dtraceserver.port=${port}`]);
+        if (server.pid === undefined) {
+            // When pid is undefined it usually means we're about to get an error.
+            return new Promise<never>((_, reject) => server.once('error', reject));
+        }
+        this.server = server;
+        this.port = port;
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        let timeout: any;
+        try {
+            await new Promise<void>((resolve, reject) => {
+                // If the server doesn't error or exit within 2 seconds, we consider that a success.
+                // That doesn't mean that the server has really started. On the frontend, the `TraceServerConnectionStatusService`
+                // will ping the port until it receives a response, which is the official measure of success.
+                timeout = setTimeout(resolve, 2000);
+                // If the server exits or errors before it outputs, consider it a failure.
```

Comment:
This comment doesn't quite apply, since we're not waiting for output, we're just waiting on a timeout.

Reply:
It's a weird one, the comment comes from the previous code but the way it seem to work is that this API is supposed to spawn the trace server, and the service telling if the server is healthy is located somewhere else and is listening to the output.

This is because this method only cares about spawning the process, but the process then can take its sweet time before actually running the server logic (e.g. initialization time past process creation).

So I kept the same logic as before:

- Spawn the server process
- Wait 2 seconds
- If the process exits for any reason during those 2s then something went wrong

Reply:
You're right that the comment is from the old code, but it was actually invalid even in the old code. When that code was being drafted, there was a time when it listened for output on the server's stdio, and that's what this comment is referring to. Before that code was merged, the output listener was removed here (invalidating the comment), and a two-part startup process was implemented on the frontend that (1) asked for the server to be started, and then (2) waited for a separate service that pings the server to declare success. It looks like the logic has been changed around a little, though, so now if the `startServer` promise resolves, the frontend tries to execute the code that requires it immediately.

https://github.com/theia-ide/theia-trace-extension/blob/64a63a19c75596bfe7f7099a6790f05f2b6e1956/theia-extensions/viewer-prototype/src/browser/trace-viewer/trace-viewer-contribution.ts#L74-L96

https://github.com/theia-ide/theia-trace-extension/blob/64a63a19c75596bfe7f7099a6790f05f2b6e1956/theia-extensions/viewer-prototype/src/browser/trace-explorer/trace-explorer-sub-widgets/trace-explorer-placeholder-widget.tsx#L82-L104

Reply:
TIL thanks for explaining how it works now, will update the comment. I'll leave it up to the Trace people to deal with that!